proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,131 @@
+   [net6.0]: Microsoft.ML (1.1.0) -> Microsoft.ML.Util (1.1.0) -> packageA (1.1.0)
+   [net472] Microsoft.ML (1.1.0) -> Microsoft.ML.Util (1.1.0) -> packageA (1.1.0)
+```
+
+- List dependency graph of a package given `package id` when there is a diamond dependency (a package is brought in by more than one path).
+
+```
+dotnet nuget why packageA
+
+Project 'projectNameA' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
+   [net6.0] Microsoft.ML (1.0.0) -> Microsoft.ML.SampleUtils (1.0.0) -> packageA (1.0.0)
+```
+
+- List dependency graph of a package given `package id` and `target framework`.
+
+```
+dotnet nuget why packageA -f net6.0
+
+Project 'projectNameA' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.0.0) -> Microsoft.ML.Util (1.0.0) -> packageA (1.0.0)
+
+Project 'projectNameB' has the following dependency graph for 'packageA'
+   [net6.0]: Microsoft.ML (1.1.0) -> Microsoft.ML.Util (1.1.0) -> packageA (1.1.0)
+```
+
+## Drawbacks 
+
+The contents of the `project.assets.json` will be helpful in understanding the dependency graph of a given package but there are a few limitations:
+
+-  There may be some packages in the file that are not part of the final graph because the NuGet restore operation downloads other packages during dependency resolution. The absence of a package id and version in the `project.assets.json` file indicates that a package was downloaded during dependency resolution but is not part of the final dependency graph.
+- Packages acquired through `PackageDownload` are not tied to the project in any way beyond acquisition. These packages are not recorded in the final graph that `project.assets.json` maintains, however they are recorded under project -> frameworks -> {frameworkName} -> downloadDependencies.
+
+## Rationale and Alternatives
+
+We could have considered a minor tweak to the existing experience of “dotnet list package --include-transitive” to provide the user with a sense of a package's dependencies. This could have been a new column next to “Resolved” called “Transitively Referenced” that had a list of the top-level packages that requested the dependency. The corresponding tracking issue can be found here: https://github.com/NuGet/Home/issues/11625. Unfortunately, this solution doesn't provide an easy way to identify the dependency graph `(starting from project -> top-level package -> transitive-dependency (1) -> ..-> transitive dependency (n))` for a particular package.
+
+![](../../meta/resources/TransitiveDependencies/TransitiveDotNetCLI.png)
+
+## Prior Art
+
+Within Visual Studio, there will be a new panel within the Visual Studio Package Manager UI named “Transitive Packages” in which all transitive packages will be displayed to the user. For the sake of not confusing the user, there will be an additional header titled “Top-level Packages”.
```

Comment:
The VS one is also lacking in the fact that it hides the transitive dependency hidden through projects. 
dotnet nuget why doesn't have that limitation.

Reply:
Could you expand on what you mean by this? Are you referring to the lack of details in the PMUI view or the solution explorer view (or potentially both)?

Reply:
In the PM UI view, transitive packages that are transitive because they come from a project are not shown in the PM UI. 

Which isn't great imo, but it was a design decision in the PM UI. 

This command wouldn't have that issue.

Reply:
Thanks for explaining, I have updated the spec to reflect that.